recirq/toric_code/optimizers.py

```diff
@@ -0,0 +1,460 @@
+    clifford = cirq.SingleQubitCliffordGate.from_unitary(unitary)
+
+    clifford_to_friendly_rep: Dict[cirq.SingleQubitCliffordGate, cirq.Gate] = {
+        cirq.SingleQubitCliffordGate.I: cirq.I,
+        cirq.SingleQubitCliffordGate.H: cirq.H,
+        cirq.SingleQubitCliffordGate.X: cirq.X,
+        cirq.SingleQubitCliffordGate.Y: cirq.Y,
+        cirq.SingleQubitCliffordGate.Z: cirq.Z,
+        cirq.SingleQubitCliffordGate.X_sqrt: cirq.X ** 0.5,
+        cirq.SingleQubitCliffordGate.Y_sqrt: cirq.Y ** 0.5,
+        cirq.SingleQubitCliffordGate.Z_sqrt: cirq.S,
+        cirq.SingleQubitCliffordGate.X_nsqrt: cirq.X ** -0.5,
+        cirq.SingleQubitCliffordGate.Y_nsqrt: cirq.Y ** -0.5,
+        cirq.SingleQubitCliffordGate.Z_nsqrt: cirq.S ** -1,
+    }
+    return clifford_to_friendly_rep.get(clifford, clifford)
+
+
+def convert_cnot_moments_to_cz_and_simplify_hadamards(
+    circuit: cirq.Circuit,
+) -> cirq.Circuit:
+    """Create a new circuit with CNOT replaced with CZ and simplified H.
+
+    We preserve the moment structure of the CNOTs and insert Hadamards as needed.
+    """
+    moment_on_deck = cirq.Moment()
+    new_circuit = cirq.Circuit()
+
+    for moment in circuit:
+        if _has_2q_gates(moment):
+            hadamard0, cz, hadamard1 = _break_up_cnots(moment)
+            new_circuit += _merge_hadamards(moment_on_deck, hadamard0)
+            new_circuit += cz
+            moment_on_deck = hadamard1
+        elif _is_exclusively_1q_gates(moment):
+            merge_moments = _merge_hadamards(moment_on_deck, moment)
+            if len(merge_moments) > 1:
+                new_circuit += merge_moments[0]
+            moment_on_deck = merge_moments[-1] or cirq.Moment()
+        else:
+            if moment_on_deck:
+                new_circuit += moment_on_deck
+                moment_on_deck = cirq.Moment()
+            new_circuit += moment
+
+    if moment_on_deck:
+        new_circuit += moment_on_deck
+
+    return new_circuit
+
+
+def defer_single_qubit_gates(circuit: cirq.Circuit) -> cirq.Circuit:
+    """Create a new circuit where some early single-qubit gates have been pushed later.
+
+    This is useful if the qubits begin the circuit in |0⟩; we leave them in the initial state
+    until right before their first entangling gates.
+
+    This is intended for a moment structure where each moment is exclusively 1q or exclusively 2q
+    gates. We do not add any additional moments, and we do not insert 1q gates into any moment
+    that contains 2q gates. We stop deferring qubit gates once we encounter a gate on that qubit
+    during a moment that contains multi-qubit gates.
+    """
+    new_circuit = copy.deepcopy(circuit)
+
+    exclusively_1q_moments: List[
+        int
+    ] = []  # We can move 1q gates in/out of these moments
+    deferrable_qubits = set(new_circuit.all_qubits())
+    qubit_to_deferred_ops: Dict[cirq.Qid, List[cirq.Operation]] = defaultdict(
+        lambda: []
+    )
```

Comment:
`defaultdict(list)` is concise and idiomatic

Reply:
Done.